### PR TITLE
Gracefully catch and handle CursorWindowAllocationException

### PR DESCRIPTION
**Pull request by djih**
This addresses: https://github.com/amplitude/Amplitude-Android/issues/42. We believe that our SDK correctly closes cursors, which would mean that these CursorWindowAllocationExceptions are caused by the app running out of memory, not something in our SDK's control. That being said, the crash still originates from our SDK, and so we should attempt to catch it and gracefully handle it, instead of crashing the app.

Our SDK uses cursors when fetching unsent events/identifies and key/value pairs. We need to catch the exception in these 3 primary use cases:

1. Fetching unsent events/identifies for upload
    * If the fetch fails, then we can simply defer the upload, and try again later maybe when the app has more free memory
2. Fetching the userId/deviceId during initialization
    * It is crucial that our SDK loads the existing userId/deviceIds during initialization, otherwise the tracked events will go to the wrong user and affect new and active user counts. If we cannot load those ids, then we cannot log any events.
    * We can set the apiKey back to null in the catch handler, as if `initialize` was never called on the SDK. All of the core SDK methods already check that an apiKey is set, and so they will just be ignored
    * The key detail here is the initialization logic is run on the background thread, so it's possible that the apiKey is set to null after a bunch of logEvents have already been scheduled (passed the apiKey check), so then we need to add an additional apiKey check inside all of the runnables
3. Fetching event metadata during logEvent
    * Whenever the SDK logs an event, it fetches some meta data like the last sequence number, increments it, sets it on the event, and then saves the new value back to the database.
    * We don't want the meta data fetching to fail, otherwise the events could end up in a weird state. What we should do is load all of the meta data into memory during initialize, and then just increment and read the values in memory. We still need to write any value updates to the database, but those don't require a Cursor object, so those should not fail

Couple extra things to note:
* I was not able to directly import the `CursorWindowAllocationException` class. I can see it in the Android Source, but I'm not able to access it directly. So instead what I have to do is catch the base class `RuntimeException` and do a string check on the message. If the exception caught is not a CursorWindowAllocationException, we need to rethrow it
* Since the entire initialization logic was moved to a background thread, the `initializeDeviceInfo` does not need to be explicitly wrapped in the `runOnLogThread`. That wrapping also prevents unit tests from running correctly since the logThread is improperly named to "main" thread in unit testing

### CODE REVIEW

File: src/com/amplitude/api/AmplitudeClient.java [hunk 1]
```diff
@@ -916,63 +945,38 @@ private long getLongvalue(String key, long defaultValue) {
...
         dbHelper.insertOrReplaceKeyLongValue(SEQUENCE_NUMBER_KEY, sequenceNumber);
         return sequenceNumber;
     }
 
     /**
-     * Internal method to get the last event time.
-     *
-     * @return the last event time
-     */
-    long getLastEventTime() {
```
Comment (curtisliu) [inline src/com/amplitude/api/AmplitudeClient.java]: My sense is we can leave these methods and just change the implementation

File: src/com/amplitude/api/AmplitudeClient.java [hunk 2]
```diff
@@ -1542,6 +1554,19 @@ public void run() {
             } catch (JSONException e) {
                 uploadingCurrently.set(false);
                 logger.e(TAG, e.toString());
+
+            // handle CursorWindowAllocationException when fetching events, defer upload
+            } catch (RuntimeException e) {
+                String message = e.getMessage();
+                if (!TextUtils.isEmpty(message) && message.startsWith("Cursor window")) {
```
Comment (curtisliu) [inline src/com/amplitude/api/AmplitudeClient.java]: Can this logic be factored into a method?

File: src/com/amplitude/api/AmplitudeClient.java [hunk 3]
```diff
@@ -225,28 +234,48 @@ public synchronized AmplitudeClient initialize(final Context context, final Stri
...
-                        client.userId = userId;
-                        client.dbHelper.insertOrReplaceKeyValue(USER_ID_KEY, userId);
-                    } else {
-                        client.userId = client.dbHelper.getValue(USER_ID_KEY);
-                    }
-                    Long optOut = client.dbHelper.getLongValue(OPT_OUT_KEY);
-                    client.optOut = optOut != null && optOut == 1;
+                if (!initialized) {
+                    try {
+                        AmplitudeClient.upgradePrefs(context);
```
Comment (curtisliu) [inline src/com/amplitude/api/AmplitudeClient.java]: verify that all steps in this try/catch block are idempotent and add a comment that they should be

File: src/com/amplitude/api/AmplitudeClient.java (hunk 2 above)
Comment (djih) [inline src/com/amplitude/api/AmplitudeClient.java]: Refactored the message checking part into a method

File: src/com/amplitude/api/AmplitudeClient.java (hunk 1 above)
Comment (djih) [inline src/com/amplitude/api/AmplitudeClient.java]: I feel like it's cleaner if we just remove them? The thing is like the calling methods will save the values to local variables that have the same name as the instance variables. Having duplicate variables + extra level of indirection is kind of messy

Comment (curtisliu) [inline src/com/amplitude/api/AmplitudeClient.java]: Yeah, I agree with you.

File: src/com/amplitude/api/AmplitudeClient.java (hunk 3 above)
Comment (djih) [inline src/com/amplitude/api/AmplitudeClient.java]: verified and commented

**Comment by djih** [comment]
Pulled the exception handling logic into the DatabaseHelper. The DatabaseHelper will throw an Amplitude `cursorWindowAllocationException` which can then be caught cleanly by the SDK client. I had to wrap db.query to facilitate mocking during unit tests to throw exceptions.

You can take another look if you want @curtisliu

**Pull request merged**